internal/cli/root.go

```diff
@@ -219,48 +223,11 @@ func (c *Client) NewCommand() *cobra.Command {
 	rootCmd.PersistentFlags().StringVar(&c.sortBy, "sort-by", "", "Sort fields for use in 'get' actions. Sort is not supported by all resources.")
 	rootCmd.PersistentFlags().StringVar(&c.sortDir, "sort-dir", "", "Sort field direction for use in 'get' actions. Sort is not supported by all resources.")
 
-	rootCmd.Version = c.Version
+	rootCmd.Version = getMetalVersion()
 	c.rootCmd = rootCmd
 	return c.rootCmd
 }
 
-// ListOptions creates a packngo.ListOptions using the includes and excludes persistent
-// flags. When not defined, the defaults given will be supplied.
-func (c *Client) ListOptions(defaultIncludes, defaultExcludes []string) *packngo.ListOptions {
```

Comment:
I believe we will want to preserve ListOptions in some way.

In Metal Go, request types that include `Include()` and `Exclude()` could be used as the basis for an interface that a new `ListOptions` would depend on.

We could take advantage of the generated methods (https://github.com/equinix-labs/metal-go/blob/main/metal/v1/api_devices.go#L831-L841). 

```go
type RequestWithOptions interface{} {
  Include([]string)
  Exclude([]string)
}
```

And create a ListOptions function that returns a function that manipulates the Request object:

```go
func (c *Client) ListOptions(defaultIncludes, defaultExcludes []string) func(RequestWithOptions ) {
  // essentially the existing code, building `[]string` for `includes` and `excludes` rather than stashing it into the ListOptions variable.
  return func(RequestWithOptions req) {
    res.Include(includes)
    res.Exclude(excludes)    
  }
}
```

Where this seems to fall apart is that we don't have spec definitions for `search`, `sort-by`, and `sort-dir`.

In order to apply the `filter` (an arbitrary set of key/value pairs: 
 `map[string]string`), we would need some function that could manipulate additional query parameters, either well-known (defined in spec) or not (arbitrary query parameters).

Thoughts, @ctreatma?

Reply:
Alternatively, we rip out the helper and do this by hand for each action 😕 until a better pattern emerges.

Reply:
The generated SDKs are extremely opinionated, which is what makes them so good at providing spec feedback.  If we start trying to bolt packngo design patterns on to metal-go, I think we'd be better off killing metal-go and keeping packngo as the supported Go SDK.

In terms of the missing parameters in the spec, we should get those updated upstream; patches are nice but it can be a pain when they're upstreamed, and these parameters are a clear case of something that should be covered in the spec (and by extension, the redoc docs).

Reply:
@ctreatma were there examples in your other 'adopt metal-go' PRs that could serve as guidance for these query parameters?

Reply:
The terraform PR makes use of `Include` in some places, such as here: https://github.com/equinix/terraform-provider-equinix/pull/291/files#diff-dd9b87265483ecd6ba13ac77846bf80816e2efc1b654fb84e897155fda5e89dfR224

The request models are chainable, so using both include and exclude parameters could look something like `obj, resp, err := req.Include(includes).Exclude(excludes).Execute()`.  Once the other parameters are in the spec, the generated code would include similar methods for requests that support those parameters as well, so you'd potentially end up with something like `req.Include(includes).Exclude(excludes).Filter(filter).Search(search).SortBy(sortBy).Execute()` (with better formatting, but there's only so much that formatting can do with that).  If I had to guess, it would actually look like a bunch of if statements so we only call, e.g. `req.Search()` if there is a non-nil value for `search`.

Reply:
Sounds good. `Filter` is a packngo-ism. It represents arbitrary query string parameters. These happen to be useful for filtering ("metro=sv").

Does metal-go have a way to tack-on query parameters (ideally from a map[string]string parameter) in a request?

Reply:
>  it would actually look like a bunch of if statements so we only call

I believe the API won't complain (or change its response) if empty strings are provided for these parameters. I agree that it would be better to not chance that.  I wonder if openapitools-generator's Go will include `omitempty` for these string fields based on some OAS3 spec.  I see https://swagger.io/docs/specification/describing-parameters/#nullable, but that "allowEmptyValue" is about permitting `?foo&bar`, not omitting those fields from the query string.